### PR TITLE
Session crud

### PR DESCRIPTION
**Pull request by buzyakabarbuzyaka**

### CODE REVIEW

File: api/cmd/api/handler/handler.go [hunk 1]
```diff
@@ -62,6 +65,9 @@ func New(baseURL string, confaCRUD *confa.CRUD, sender *email.Sender, producer P
 	r.POST("/confa/v1/confas", h.createConfa)
 	r.GET("/confa/v1/confas/:confa_id", h.confa)
 
+	r.POST("/confa/v1/sessions", h.createSession)
+	r.GET("/confa/v1/sessions", h.session) // unmarshal session key from request
```
Comment (buzyakabarbuzyaka) [inline api/cmd/api/handler/handler.go]: Base64 can contain '/' symbol. It destroys routing.

File: api/cmd/api/handler/http.go [hunk 2]
```diff
@@ -70,6 +69,66 @@ func (h *Handler) confa(w http.ResponseWriter, r *http.Request, ps httprouter.Pa
...
+		return
+	}
+
+	var request session.Session
+	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
+		_ = api.BadRequest(api.CodeMalformedRequest, err.Error()).Write(ctx, w)
+		return
+	}
+
+	sess, err := h.sessionCRUD.Create(ctx, userID, request)
```
Comment (buzyakabarbuzyaka) [inline api/cmd/api/handler/http.go]: This route was made to test fetch logics in session witch might be unnecessary.

File: api/internal/user/session/session.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	ErrDuplicatedEntry  = errors.New("duplicated entry")
+)
+
+type Session struct {
+	Key       string    `json:"key"`
+	Owner     uuid.UUID `json:"owner"`
+	CreatedAt time.Time `json:"createdAt"`
+}
+
+const byteKeyLength = 96 // you need 4*(n/3) chars to represent n bytes
```
Comment (aromancev) [inline api/internal/user/session/session.go]:
```suggestion
const keyLength = 96 // you need 4*(n/3) chars to represent n bytes
```

File: api/internal/user/session/session.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	_, err := rand.Read(b)
+	if err != nil {
+		panic(err)
+	}
+
+	key := base64.StdEncoding.EncodeToString(b)
+	return key
+}
+
+func (s Session) GenerateKey() (Session, error) {
```
Comment (aromancev) [inline api/internal/user/session/session.go]:
```suggestion
func (s Session) New() Session {
```

`GenerateKey`  sounds like it should return some kind of key. But in fact it returns `Session`. Let's make it a constructor instead.

File: api/internal/user/session/session.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+type Session struct {
+	Key       string    `json:"key"`
+	Owner     uuid.UUID `json:"owner"`
+	CreatedAt time.Time `json:"createdAt"`
+}
+
+const byteKeyLength = 96 // you need 4*(n/3) chars to represent n bytes
+func generateKey() string {
```
Comment (aromancev) [inline api/internal/user/session/session.go]: If it's only used inside `GenerateKey` maybe we should just copy-paste this code there?

File: api/internal/user/session/session.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	if s.Key != "" {
+		return s, errors.New("key is not empty")
+	}
+
+	key := generateKey()
+	s.Key = key
+	return s, nil
+}
+
+func (s Session) Validate() error {
```
Comment (aromancev) [inline api/internal/user/session/session.go:37]: Also need to check if key is not empty.

**Review by aromancev — CHANGES_REQUESTED** [review]

File: api/internal/auth/auth.go [hunk 7]
```diff
@@ -101,3 +102,17 @@ func Authenticate(r *http.Request) (uuid.UUID, error) {
 	id, _ := uuid.Parse("28164069-5ec3-405b-a9cc-641cf29588ed") //todo: Unhardcode this. ONLY FOR TESTING
 	return id, nil
 }
+
+func TokenHelper(rawToken string) (string, error) {
```
Comment (aromancev) [inline api/internal/auth/auth.go]:
```suggestion
func Bearer(r *http.Request) (string, error) {
```

File: api/internal/auth/auth.go [hunk 8]
```diff
@@ -101,3 +102,17 @@ func Authenticate(r *http.Request) (uuid.UUID, error) {
...
+	authArray := strings.Split(rawToken, " ")
+	if len(authArray) < 2 {
+		return "", errors.New("header format error")
```
Comment (aromancev) [inline api/internal/auth/auth.go]:
```suggestion
		return "", errors.New("wrong header format")
```

File: api/internal/auth/auth.go [hunk 9]
```diff
@@ -101,3 +102,17 @@ func Authenticate(r *http.Request) (uuid.UUID, error) {
...
+	}
+
+	bearer, token := authArray[0], authArray[1]
+	if bearer != "Bearer" {
+		return "", errors.New("header type format error")
```
Comment (aromancev) [inline api/internal/auth/auth.go]:
```suggestion
		return "", errors.New("wrong type")
```

File: api/internal/user/session/session.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	b := make([]byte, keyLength)
+	_, err := rand.Read(b)
+	if err != nil {
+		panic(err)
+	}
+	key := base64.StdEncoding.EncodeToString(b)
+
+	s := Session{}
+	s.Key = key
+	return s
```
Comment (aromancev) [inline api/internal/user/session/session.go]:
```suggestion
	return Session{Key: key}
```

**Review by aromancev — APPROVED** [review]

**Pull request merged**